File: 2516/Solution.java
class Solution {
    public int takeCharacters(String s, int k) {
        int[] a = new int[3];
        for(char c :  s.toCharArray()) a[c - 'a']++;

        if(a[0] < k || a[1] < k || a[2] < k) return -1;

        int res = Integer.MAX_VALUE;
        for(int l = 0, r = 0; r < s.length(); r++){
            a[s.charAt(r) - 'a']--;

            while(l < s.length() && a[0] < k || a[1] < k || a[2] < k) {
                a[s.charAt(l++) - 'a']++;
            }

            res = Math.min(res, s.length() - (r - l + 1));
        }
        return res;
    }
}

File: 2516/Solution.go
func takeCharacters(s string, k int) int {
    a := make([]int, 3)
    res, n := math.MaxInt, len(s)
    for _, c := range s {
        a[c - 'a']++
    }

    if a[0] < k || a[1] < k || a[2] < k {
        return -1
    }
    
    for l, r := 0, 0; r < n; r++ {
        a[s[r] - 'a']--

        for l < n && a[0] < k || a[1] < k || a[2] < k {
            a[s[l] - 'a']++
            l++
        }

        res = min(res, n - (r - l + 1))
    }

    return res
}
